### ISSUE TITLE
cookie_secure setting looks unnecessary

### ISSUE THREAD

**Issue by ekini** [silverstripe/silverstripe-framework#9346]
## Affected Version

at least 4.4

## Description

I would expect to see the "Secure" cookie flag set by default for https connections, however without `cookie_secure: true` in the config it doesn't happen. It's not described in [the documentation](https://docs.silverstripe.org/en/4/developer_guides/cookies_and_sessions/sessions/#secure-session-cookie) and the only way to know it is to look at [the code](https://github.com/silverstripe/silverstripe-framework/blob/b002ef1171f5133735eafbc63d937eeb586c2e01/src/Control/Session.php#L293)

For example, instead of
```
Set-Cookie: PHPSESSID=<sessionid>; expires=Mon, 29-Oct-2018 21:38:08 GMT; Max-Age=3600; path=/; HttpOnly
```
I expect to see
```
Set-Cookie: PHPSESSID=<sessionid>; expires=Mon, 29-Oct-2018 21:38:08 GMT; Max-Age=3600; path=/; Secure; HttpOnly
```

If I set `cookie_secure: true`, it renames the default cookie name to `SECSESSID`, which as a non-standard name can be accidentally exposed, for example in Sentry:
![2019-12-09 07 52 10](https://user-images.githubusercontent.com/1527663/70394303-de76e700-1a58-11ea-94c5-dacd7dd8ef1b.jpg)

So to make it correct I have to add another setting that renames the cookie back:
```yaml
SilverStripe\Control\Session:
  cookie_secure: true
  cookie_name_secure: 'PHPSESSID'
```

The above mechanism of renaming the cookie claims it adds another layer of security
> This uses the session_name SECSESSID for https connections instead of the default PHPSESSID. Doing so adds an extra layer of security to your session cookie since you no longer share http and https sessions.

however, with "Secure" flag set, cookies won't be shared between http/https without any extra configuration / code.

## Steps to Reproduce

As an example, look at https://www.silverstripe.org/Security/login?BackURL=%2Fadmin%2Fpages which has https but doesn't have "Secure" flag on the cookie.

## Proposed solution

Deprecate the setting, change [the line](https://github.com/silverstripe/silverstripe-framework/blob/b002ef1171f5133735eafbc63d937eeb586c2e01/src/Control/Session.php#L293)
to
```php
$secure = Director::is_https($request)
```
It will make the default installation more secure and remove the code that weakens security (by renaming the session cookie).

**Comment by maxime-rainville** [comment]
I'm thinking maybe it's time for us to adopt an HTTPS-first approach and treat HTTP as the exception.

At first glance and not having looked at the underlying code, the suggestion sounds sensible. Although, I'm not sure if we can implement it in the 4 branch without breaking things for some users.

**Comment by dnsl48** [comment]
Agree, that would be a breaking change for the existing installations.
However, I believe it's worth doing for the next major release (Silverstripe CMS 5).
I'd also say it's rather `impact/high` since it affects all https installations.

**Comment by ekini** [comment]
I think there are two parts here:
1. One is related to "Secure" flag to the session cookie. Can we make it non-dependable on `cookie_secure: true`? I can't think of any example when it might break something for non-HTTPS websites, but it needs more thinking.
2. Deprecation of `cookie_secure` setting. If people depend on the cookie name `SECSESSID` , renaming it back will definitely break something. However, it would be nice to deprecate it now, if possible, and add the following to the documentation of v4 (until the point one has been implemented):
```yaml 
cookie_secure: true # sets Secure flag on the cookie
cookie_name_secure: 'PHPSESSID' # leaves the default name because many tools hide PHPSESSID cookie value but don't do it for custom names
```

It's just because right now default installations are insecure: without `cookie_secure: true` the Secure flag is not set, so the cookie might leak, and with `cookie_secure: true` the cookie is renamed to a custom name, and the value might leak to logs / error tracking software.

**Comment by dnsl48** [comment]
> One is related to "Secure" flag to the session cookie. Can we make it non-dependable on cookie_secure: true? 

I think the whole point of `cookie_secure` parameter is to control this flag. I'm not sure if we should make session cookie independent, but it's definitely worth making it `true` by default (IMO).

A different story is the cookie name. I believe we may want to make it `PHPSESSID` by default, but we should be careful with that. The problem we might face is that PHP usually controls that cookie record on its own and exposes some settings through `php.ini`. I think we should learn better how exactly it works across the different PHP versions before we make this change.

**Comment by sunnysideup** [comment]
BUMP. To me, it it would be really good that a vanilla install of Silverstripe is super secure, and then you can make it less secure, if needed.

**Comment by GuySartorelli** [comment]
Cookies will be secure by default in CMS 6.
Closing this in favour of https://github.com/silverstripe/silverstripe-framework/issues/11669 which handles the question about the session cookie name.
